byron-adapter/src/Ouroboros/Byron/Proxy/Index.hs

```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE GADTs #-}
+
+module Ouroboros.Byron.Proxy.Index where
+
+-- cardano-sl crypto apparently gives no way to get an AbstractHash from
+-- a ByteString without re-hashing it...
+import Crypto.Hash (Digest, HashAlgorithm, digestFromByteString)
+import Data.ByteString (ByteString)
+import Data.ByteArray (convert)
+import Database.SQLite.Simple as Sql
+import System.Directory (doesFileExist)
+
+import Pos.Chain.Block (HeaderHash)
+import Pos.Crypto.Hashing (AbstractHash (..))
+
+import Ouroboros.Storage.ImmutableDB.Types
+
+digestFromByteString_partial :: HashAlgorithm a => ByteString -> Digest a
+digestFromByteString_partial bs =
+  let Just digest = digestFromByteString bs in digest
+
+newtype OwnHash = OwnHash
+  { getOwnHash :: HeaderHash
+  }
+
+newtype ChildHash = ChildHash
+  { getChildHash :: Maybe HeaderHash
+  }
+
+data IndexPoint next where
+  Tip    :: IndexPoint OwnHash
+  ByHash :: HeaderHash -> IndexPoint ChildHash
+
+-- | Given an `IndexPoint`, look it up in the index and give the forward
+-- link (determined by the `IndexPoint`s parameter) and the `EpochSlot` for
+-- the entry at the index point.
+--
+-- For the tip you need to get a header hash (of the tip)
+-- For others you need to get a header hash, but of its child.
+data Index m = Index
+  { indexAt :: forall datum . IndexPoint datum -> m (Maybe (datum, EpochSlot))
+  , updateTip :: HeaderHash -> EpochSlot -> m ()
+  }
+
+-- Practical details:
+-- - We'll use rocksdb, since there's precedent, and a key/value is more
+--   appropriate than relational (how would we keep the tip in a relational?)
+-- - Oh but never mind, in a relational we can easily get the tip by ordering
+--   the (headerhash, epochslot) relation.
```

Comment:
I like the discourse!

Reply:
Hah, normally I remove all traces of me talking to myself.

Reply:
You did in subsequent patches :-)